sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuRegExpReplaceMeta.scala

```diff
@@ -20,46 +20,90 @@ import org.apache.spark.sql.rapids.{GpuRegExpReplace, GpuRegExpReplaceWithBackre
 
-  private var javaPattern: Option[String] = None
-  private var cudfPattern: Option[String] = None
-  private var replacement: Option[String] = None
-  private var canUseGpuStringReplace = false
-  private var containsBackref: Boolean = false
+  private var action: Option[GpuRegExpReplaceAction] = None
 
   override def tagExprForGpu(): Unit = {
     GpuRegExpUtils.tagForRegExpEnabled(this)
-    replacement = expr.rep match {
-      case Literal(s: UTF8String, DataTypes.StringType) if s != null => Some(s.toString)
-      case _ => None
+    val replacement: String = expr.rep match {
+      case Literal(s: UTF8String, DataTypes.StringType) if s != null => s.toString
+      case _ =>
+        willNotWorkOnGpu(s"only non-null literal strings are supported on GPU")
+        return
     }
 
     expr.regexp match {
       case Literal(s: UTF8String, DataTypes.StringType) if s != null =>
-        if (GpuOverrides.isSupportedStringReplacePattern(expr.regexp)) {
-          canUseGpuStringReplace = true
-        } else {
-          try {
-            javaPattern = Some(s.toString())
-            val (pat, repl) = 
-                new CudfRegexTranspiler(RegexReplaceMode).getTranspiledAST(s.toString, replacement)
-            GpuRegExpUtils.validateRegExpComplexity(this, pat)
-            cudfPattern = Some(pat.toRegexString)
-            repl.map { r => GpuRegExpUtils.backrefConversion(r.toRegexString) }.foreach {
+        try {
+          val (pattern, repl) = new CudfRegexTranspiler(RegexReplaceMode)
+              .getTranspiledAST(s.toString, Some(replacement))
+          GpuRegExpUtils.validateRegExpComplexity(this, pattern)
+          val maybeBackref = repl.map { r => GpuRegExpUtils.backrefConversion(r.toRegexString) }
+          val hasBackRef = maybeBackref.exists(_._1)
+
+          pattern match {
+            case RegexChoice(a, b) if !hasBackRef =>
+              // special handling for "AB|CD" where we can use string replace instead of
+              // regexp replace for improved performance. We can only perform this optimization
+              // if the following conditions are met:
+              //
+              // - Both strings must be supported by string_replace, so cannot contain regexp chars
+              // - There must be no back-references
+              // - The second string must not overlap with the replacement string because this
+              //   could cause incorrect results. For example, if we replace "AB|CD" with "C" for
+              //   the input "ABD", we would first replace "AB" with "C", resulting in "CD", which
+              //   would now match the second string, and the original string would not have
+              //   matched.
+              val str1 = a.toRegexString
+              val str2 = b.toRegexString
+              //TODO this check is probably too broad at the moment and could be refined
+              val overlap = str2.contains(replacement)
```

Comment:
I now realize that this check is not sufficient, and I am no longer sure that this approach can even work.

If we have an input string of "cabd" and our pattern is "ab|cd" and replacement is an empty string, then the first replace will result in "cd" and the second replace will result in "", but the expected result is "cd".

Reply:
I'm exploring a solution involving using a placeholder when performing the first string_replace. This will result in a much more complex expression, so the performance may not be good enough.

The approach I am trying will be something like this:

```
if (string_contains(input, placeholder)) { 
  // fallback to regexp_replace
  regexp_replace(...)
} else {
  repl1 = string_replace(input, string1, placeholder)
  repl2 = string_replace(repl1, string2, replacement_string)
  string_replace(repl2, placeholder, replacement_string)
}
```

Reply:
The multi-replace I pointed out might be the only solution then.

Reply:
Ya, I just looked at the implementation for the mult-replace and it is clearly not optimized for long strings.